### ISSUE TITLE
instanceof check for interfaces is static, not dynamic

### ISSUE THREAD

**Issue by CountBleck** [AssemblyScript/assemblyscript#2577]
Compare and contrast the following two snippets of code: [`<Object> instanceof <class>`](https://www.assemblyscript.org/editor#IyFvcHRpbWl6ZT1ub25lJnJ1bnRpbWU9aW5jcmVtZW50YWwKY2xhc3MgRm9vIHt9DQpjbGFzcyBCYXIgZXh0ZW5kcyBGb28ge30NCg0KY29uc3QgdGVzdCA9IGNoYW5nZXR5cGU8QmFyPigwKQ0KDQpleHBvcnQgZnVuY3Rpb24gaGV5KCk6IGJvb2wgew0KICByZXR1cm4gdGVzdCBpbnN0YW5jZW9mIEZvbw0KfQ0KDQpleHBvcnQgZnVuY3Rpb24gaGl5YSgpOiBib29sIHsNCiAgcmV0dXJuIHRlc3QgYXMgT2JqZWN0IGluc3RhbmNlb2YgRm9vDQp9CiMhaHRtbAo=) and [`<Object> instanceof <interface>`](https://www.assemblyscript.org/editor#IyFvcHRpbWl6ZT1ub25lJnJ1bnRpbWU9aW5jcmVtZW50YWwKY2xhc3MgRm9vIHt9DQpjbGFzcyBCYXIgZXh0ZW5kcyBGb28ge30NCg0KY29uc3QgdGVzdCA9IGNoYW5nZXR5cGU8QmFyPigwKQ0KDQpleHBvcnQgZnVuY3Rpb24gaGV5KCk6IGJvb2wgew0KICByZXR1cm4gdGVzdCBpbnN0YW5jZW9mIEZvbw0KfQ0KDQpleHBvcnQgZnVuY3Rpb24gaGl5YSgpOiBib29sIHsNCiAgcmV0dXJuIHRlc3QgYXMgT2JqZWN0IGluc3RhbmNlb2YgRm9vDQp9CiMhaHRtbAo=)

The former snippet works as expected: the `hiya` export calls an `instanceof` helper, which then checks the ID at runtime.
The latter snippet does not work as expected: the `hiya` export returns `false`, because the `instanceof` is incorrectly checked at compile-time.

To fix this bug, the latter snippet should also perform the check at runtime. In other words, `<type> instanceof <interface>` with a `<type>` that doesn't implement `<interface>` should be checked at runtime, because its subclasses can implement it.

**Comment by dcodeIO** [comment]
Indeed, `<Object> instanceof <Interface>` should call `~instanceof|<interface>`. The code to generate such helpers for interfaces exists, but as you found it's somehow not called.

**Comment by dcodeIO** [comment]
Looking at the snippets again, the second one reads

```ts
class Foo {}
class Bar extends Foo {}

const test = changetype<Bar>(0)

export function hey(): bool {
  return test instanceof Foo
}

export function hiya(): bool {
  return test as Object instanceof Foo
}
```

There, what happens is that `test` is actually `null` (though `changetype`), which is why `hiya` returns `false` (`null` is not an instance of anything).

```wat
  global.get $empty/test
  local.tee $0
  i32.eqz
  if (result i32) ;; here
   i32.const 0
  else
   local.get $0
   call $~instanceof|empty/Foo
  end
```

Do you have an updated one that shows the interface issue?

**Comment by dcodeIO** [comment]
Fixed by #2588
